### PR TITLE
Reduce memory usage in DEX files by caching the getstr calls ##bin

### PR DESCRIPTION
**Pull request by radare**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Benchmark results**

```
type    top    vsz     rss      time
------------------------------------
calloc  339M   4732640 365452   48s
pvector 338M   4731012 371484   48s // XXX fails somehow with trash indexes
htup    338M   4731556 371620   58s
nocache 350M   4742936 376156   52s
```

nocache -> calloc = 8% faster and 3% less memory usage

IMAGE 2020-04-27 20:02:31
IMAGE 2020-04-27 20:02:35

### CODE REVIEW

**Comment by radare** [comment]
The memory allocated is leaking, but it was actually leaking more before.
On performance i dont see much difference in times. this pr introduced 2 optimizations that result in less cpu usage but not really noticeable, but worth cleaning up imho:

* when calling getstr() twice with the same idx the previously allocated string is returned.
* Reduce the calls to RBuf.read_le32_at, with just one rbuf_read call and many r_read_le32. to avoid the amount of internal work in RBuffer with IO backend (seek, etc..) which adds many syscalls taht are not needed.

Ill rewrite the hashtable with just an RPVector which is exactly what we need in here

File: libr/bin/format/dex/dex.c [hunk 1]
```diff
@@ -14,6 +14,12 @@ char* r_bin_dex_get_version(RBinDexObj *bin) {
 	return NULL;
 }
 
+// XXX this is never called
```
Comment (XVilka) [inline libr/bin/format/dex/dex.c:17]: Why not to remove it then, if its never called?

File: libr/bin/p/bin_dex.c [hunk 2]
```diff
@@ -1620,7 +1634,7 @@ static void parse_class(RBinFile *bf, RBinDexClass *c, int class_index, int *met
 			rbin->cb_printf ("  source_file_idx   : %d (%s)\n\n",
 					 c->source_file, source_file);
 		}
-		free (source_file);
+		// free (source_file);
```
Comment (XVilka) [inline libr/bin/p/bin_dex.c:1637]: Why not to remove free then?

File: libr/bin/p/bin_dex.c [hunk 3]
```diff
@@ -1551,9 +1565,9 @@ static void parse_class(RBinFile *bf, RBinDexClass *c, int class_index, int *met
 			if (t > 0 && t < dex->header.types_size ) {
 				int tid = dex->types[t].descriptor_id;
 				if (dexdump) {
-					char *cn = getstr (dex, tid);
+					const char *cn = getstr (dex, tid);
 					rbin->cb_printf ("    #%d              : '%s'\n", z, cn);
-					free (cn);
+					// free (cn);
```
Comment (XVilka) [inline libr/bin/p/bin_dex.c:1570]: Same here, why not to remove free?

File: libr/bin/p/bin_dex.c [hunk 4]
```diff
@@ -1372,7 +1385,7 @@ static void parse_dex_class_method(RBinFile *bf, RBinDexClass *c, RBinClass *cls
 									s,
 									handler_addr);
 							}
-							free (s);
+							s = NULL; // free (s);
```
Comment (XVilka) [inline libr/bin/p/bin_dex.c:1388]: memory leak?

File: libr/bin/p/bin_dex.c [hunk 5]
```diff
@@ -78,31 +78,43 @@ static ut64 dex_field_offset(RBinDexObj *bin, int fid) {
...
 	if (!len || len >= dex->size) {
 		return NULL;
 	}
 	ut8 *ptr = malloc (len + 1);
-	if (!ptr) {
-		return NULL;
+	if (ptr) {
+		r_buf_read_at (dex->b, string_index + uleblen, ptr, len);
+		ptr[len] = 0;
+		if (dex->vec_strings.v.len) {
```
Comment (ret2libc) [inline libr/bin/p/bin_dex.c:109]: i don't understand why you didn't use `HtUP`. I honestly don't understand this code. So... if you have more than one element in the vector, you initialize it, right? Then, few lines below, if you have more than one element in the vector, you insert `ptr` ad `idx`.

What happens if the vector is empty and then you try to insert an element in the RPVector at index=10?  IMHO, an HtUP perfectly fits what you need. I don't see why not using it.

File: libr/bin/format/dex/dex.c (hunk 1 above)
Comment (radare) [reply libr/bin/format/dex/dex.c:17]: because its a not for my future self to fix the memory leaks that have been there for ever

File: libr/bin/p/bin_dex.c (hunk 5 above)
Comment (radare) [reply libr/bin/p/bin_dex.c:109]: it happens that it returns null, htup is unnecessary overengineering when we just need an array

File: libr/bin/p/bin_dex.c (hunk 4 above)
Comment (radare) [reply libr/bin/p/bin_dex.c]: nope, because its not freed until the the rbindex obj is freed

File: libr/bin/p/bin_dex.c (hunk 3 above)
Comment (radare) [reply libr/bin/p/bin_dex.c]: same as above

Comment (radare) [reply libr/bin/p/bin_dex.c]: getstr returns const char*, not owned, so not leaking

File: libr/bin/p/bin_dex.c (hunk 2 above)
Comment (radare) [reply libr/bin/p/bin_dex.c]: ok will do now

File: libr/bin/format/dex/dex.c (hunk 1 above)
Comment (radare) [reply libr/bin/format/dex/dex.c:17]: ill fix add the callback in all the plugins to free the rbinobj instance and all the related leaks in a separate PR

**Comment by ret2libc** [comment]
@radare I'm not sure this works. Can you provide a testcase where the new rpvector is used? I put breakpoints on the lines where the rpvector is used and it never breaks (I tried dex/test.dex, dex/hello.dex, java/Main.dex)

Also, why a cache reduce the memory usage? Doesn't it increase it?

**Comment by radare** [comment]
all the DEX tests in the testsuite use this code. if the code wasnt correct it will crash or fail

**Comment by radare** [comment]
cache improves memory usage because the string pool is queries 999999 times, 800000 of them for the same strings. as long as you cant read or test the PR ill explain again:

before:

* getstr = returns NEW string in the heap with strdup

after:

* getstr = caches the index -> string and just returns a weak pointer, the heap pointer is owned by the rpvector, so its not freed until the RBinObject is freed, which right now it doesnt happen because its a bug in the RBin api design that i plan to fix in another pr when this gets merged

File: libr/util/vector.c [hunk 6]
```diff
@@ -143,10 +143,13 @@ R_API void r_vector_remove_at(RVector *vec, size_t index, void *into) {
 
 R_API void *r_vector_insert(RVector *vec, size_t index, void *x) {
 	r_return_val_if_fail (vec, NULL);
-	if (vec->len >= vec->capacity) {
+	if (R_MAX (vec->len, index) >= vec->capacity) {
```
Comment (ret2libc) [inline libr/util/vector.c:146]: why this? index >= vec->capacity and you use NEXT_VECTOR_CAPACITY you may end up anyway with a vector that is not big enough for index.

Comment (radare) [reply libr/util/vector.c:146]: damn this shouldnt be commited, this is wrong. ill remove it and fix the CI i pushed the code because i had to run afk. and ill open an issue for all the problems i found in RVector and RPVector

**Comment by ret2libc** [comment]
So what did you choose here? I'm ok with calloc, but we should ensure strings size cannot be too big.

File: libr/include/r_vector.h [hunk 7]
```diff
@@ -158,6 +158,7 @@ R_API void *r_vector_shrink(RVector *vec);
 // RPVector
 
 R_API void r_pvector_init(RPVector *vec, RPVectorFree free);
+R_API void r_pvector_fini(RPVector *vec);
```
Comment (radare) [inline libr/include/r_vector.h:161]: this is unrelated

File: libr/bin/p/bin_dex.c [hunk 8]
```diff
@@ -1988,11 +1999,11 @@ static RList *sections(RBinFile *bf) {
 			ptr->size = ptr->vaddr; // fix size
 		}
 		ptr->vsize = ptr->size;
+		// Commenting this line speedups loading from 4.5s to 3s
```
Comment (radare) [inline libr/bin/p/bin_dex.c:2002]: @thestr4ng3r is on this

File: libr/include/r_vector.h (hunk 7 above)
Comment (radare) [reply libr/include/r_vector.h:161]: but needed because it was not defined :(

**Comment by radare** [comment]
fixed the lgtm

File: libr/bin/p/bin_dex.c [hunk 9]
```diff
@@ -76,33 +76,42 @@ static ut64 dex_field_offset(RBinDexObj *bin, int fid) {
...
-	ut64 len;
-	int uleblen;
-	if (!dex || idx < 0 || idx >= dex->header.strings_size || !dex->strings) {
+	if (idx < 0 || idx >= dex->header.strings_size || !dex->strings) {
 		return NULL;
 	}
-	if (dex->strings[idx] >= dex->size) {
+	if (dex->cal_strings) {
+		const char *p = dex->cal_strings[idx];
+		if (p && *p) {
```
Comment (ret2libc) [inline libr/bin/p/bin_dex.c:86]: `R_STR_ISNOTEMPTY`

File: libr/bin/format/dex/dex.c [hunk 10]
```diff
@@ -63,34 +74,36 @@ RBinDexObj *r_bin_dex_new_buf(RBuffer *buf) {
...
 	/* strings */
 	#define STRINGS_SIZE ((dexhdr->strings_size + 1) * sizeof (ut32))
-	bin->strings = (ut32 *) calloc (dexhdr->strings_size + 1, sizeof (ut32));
-	if (!bin->strings) {
+	if (dexhdr->strings_size > bin->size) {
 		goto fail;
 	}
-	if (dexhdr->strings_size > bin->size) {
-		free (bin->strings);
+	bin->strings = (ut32 *) calloc (dexhdr->strings_size + 1, sizeof (ut32));
```
[the remark this answers is not in the document]
Comment (radare) [reply libr/bin/format/dex/dex.c:80]: done

File: libr/bin/format/dex/dex.c [hunk 11]
```diff
@@ -134,13 +148,10 @@ RBinDexObj *r_bin_dex_new_buf(RBuffer *buf) {
 	}
 
 	/* types */
-	int types_size = dexhdr->types_size * sizeof (struct dex_type_t);
+	size_t types_size = dexhdr->types_size * sizeof (struct dex_type_t);
```
[the remark this answers is not in the document]
Comment (radare) [reply libr/bin/format/dex/dex.c:151]: yes, but not in this pr because its unrelated

**Comment by radare** [comment]
as i said 3 times, the leak introduced here by allocating cal_strings will be fixed in another PR because it affects ALL the bin plugins. i even filled an issue with more details

**Comment by ret2libc** [comment]
> as i said 3 times, the leak introduced here by allocating cal_strings will be fixed in another PR because it affects ALL the bin plugins. i even filled an issue with more details

Ok :/ then there is no need for that function. but ok.

Fix CI and then I can approve.

File: libr/bin/p/bin_dex.c (hunk 8 above)
Comment (thestr4ng3r) [reply libr/bin/p/bin_dex.c:2000]: wat am I?

File: libr/bin/format/dex/dex.c [hunk 12]
```diff
@@ -63,34 +74,36 @@ RBinDexObj *r_bin_dex_new_buf(RBuffer *buf) {
...
 	/* classes */
 	// TODO: not sure about if that is needed
-	int classes_size = dexhdr->class_size * DEX_CLASS_SIZE;
+	size_t classes_size = dexhdr->class_size * DEX_CLASS_SIZE;
 	if (dexhdr->class_offset + classes_size >= bin->size) {
-		classes_size = bin->size - dexhdr->class_offset;
-	}
-	if (classes_size < 0) {
-		classes_size = 0;
+		if (dexhdr->class_offset < bin->size) {
```
[the remark this answers is not in the document]
Comment (radare) [reply libr/bin/format/dex/dex.c:97]: unrelated to this pr

**Comment by radare** [comment]
let's wait for the CI
